### PR TITLE
Search time improvements

### LINKED ISSUE #60: Improve search responsiveness
We have a few options to improve the search speed.

Possible approachs include: 
* distro specific search
* prefetching indices and local caching of indices, possibly local storage.
* relayout site to load search elements earlier
* ...

### PR DESCRIPTION
**Pull request by hidmic**
Closes #60. Connected to #28. This pull requests performs basic `lunr.js` index sharding. 

An earlier commit witnesses a previous attempt to use browser's `sessionStorage` for index persistence. This approach was shortly dismissed, given our index exceeds the storage quota (5-10 MB tops depending on the browser, while our index uncompressed weighs ~15MB) and that client-side compression exhibited too much overhead.

### CODE REVIEW

**Comment by hidmic** [comment]
I'll leave rosindex building locally to see how it performs when indexes are the largest. It's taking a long while to fetch everything.

**Comment by hidmic** [comment]
Alright, now it seems to work properly. @chapulina would you mind deploying it? I haven't been able to fully build it (it takes hours and hours for me).

**Comment by chapulina** [comment]
@hidmic , this branch has been deployed to https://index.ros.org/ as of https://github.com/ros-infrastructure/index.ros.org/commit/3d8e14465231f87bf11f5e79ded1125e806fdfe5 . It may take a while until the most recent version is actually served by github pages, I'm not sure.

**Comment by hidmic** [comment]
Alright, one last improvement @chapulina. Responsiveness was suffering because we were pulling all shards at the same time. I've split them into smaller, serialized chunks. It seems to be working better now :).

**Comment by chapulina** [comment]
Latest version deployed: https://github.com/ros-infrastructure/index.ros.org/commit/4491bc7ed061f6100901bf112e998d6b36b5d200

**Comment by hidmic** [comment]
Now it's ready to review, thanks for the help @chapulina !

File: _layouts/package_search.html [hunk 1]
```diff
@@ -46,41 +48,41 @@ <h3>Loading Index...</h3>
...
       <tr>
-        <th style="text-align: center;"><span class="glyphicon glyphicon-flash"></span></th>
-        <th style="text-align: center;"><span class="glyphicon glyphicon-file"></span></th>
-        <th style="text-align: center;"><span class="glyphicon glyphicon-time"></span></th>
-        <th>Distro</th>
-        <th>Name</th>
-        <th>Repo</th>
+        <th class="col-xs-1" style="text-align: center;"><span class="glyphicon glyphicon-flash"></span></th>
+        <th class="col-xs-1" style="text-align: center;"><span class="glyphicon glyphicon-file"></span></th>
+        <th class="col-xs-1" style="text-align: center;"><span class="glyphicon glyphicon-time"></span></th>
```
Comment (chapulina) [inline _layouts/package_search.html:53]: Tangential to this PR, but these could also use some tooltips like the ones on https://github.com/ros2/rosindex/pull/68

**Review by chapulina — COMMENTED** [review]
The search works well for me :+1: 

Full disclosure: my observations may not be this PR's fault, this is just the 1st time I'm reviewing this. That said, some things I noticed:

* Searching for [title:cppcheck](https://index.ros.org/search/?q=title%3Acppcheck) leaves the UI in a bad state and prints an error to the console. We could handle unrecognized fields better.
* Navigating to [/search](https://index.ros.org/search/) seems to perform a search (for an empty string?). We could speed up loading this page if we didn't perform any searches when there's no query.
* From `/search`, the search results are fluidly updated without a progress bar as the query is typed. If the user presses `enter`, the page is reloaded and the search bar shows up again. I think the nice thing that happens when pressing enter is that the address bar gets updated with `q=search_term`. It would make the interaction more seamless if pressing enter didn't trigger a refresh, but just updated the address bar.

**Comment by hidmic** [comment]
@chapulina thanks for the review! All your observations are perfectly valid.

> Searching for title:cppcheck leaves the UI in a bad state..

I checked and it looks like `lunr.js` throws if an unknown field is given. At the very least we should catch it.

> Navigating to /search seems to perform a search

Truth is that the progress bar is a mock up shown while loading the index. Actual searches take a few milliseconds and no animation is triggered for those. And we don't currently handle empty queries in any special way. We could hide the entire panel. We could also just change the "Searching packages..." label to a "Loading index..." one :)

> It would make the interaction more seamless if pressing enter didn't trigger a refresh

Sure, that makes sense. The search box on the top right corner would still behave the same (i.e. refreshing), as that one is common to the whole site.

> use some tooltips

Not only that, we don't even have sorting capabilities.

In any case, I'd rather defer resolution to future PRs. This one's been open for a while now. What do you think?

**Comment by hidmic** [comment]
@chapulina if you agree, I'll merge and open up a few issues to track your observations.

File: _layouts/package_search.html (hunk 1 above)
Comment (chapulina) [reply _layouts/package_search.html:53]: Ticketed https://github.com/ros2/rosindex/issues/70

**Comment by chapulina** [comment]
Sounds good, I ticketed a few issues which may be tackled later.

**Review by chapulina — APPROVED** [review]

**Pull request merged**